File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.py
from heapq import *
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        head = ListNode()
        tail = head
        heap = []
        heapify(heap)
        
        for node in lists:
            if node is not None:
                heappush(heap, (node.val, id(node), node))
        
        while heap:
            val, idx, node = heappop(heap)
            tail.next = node
            tail = node
            
            if tail.next is not None:
                heappush(heap, (tail.next.val, id(tail.next), tail.next))
        return head.next

File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<Integer> minheap = new PriorityQueue<Integer>();
        for(ListNode node: lists) {
            while(node != null) {
                minheap.add(node.val);
                node = node.next;
            }
        }
        ListNode head = new ListNode(-1);
        ListNode tail = head;
        while(!minheap.isEmpty()) {
            tail.next = new ListNode(minheap.remove());
            tail = tail.next;
        }
        return head.next;
    }
}